### ISSUE TITLE
Failed block validation during IDB followed by leveldb corruption

### ISSUE THREAD

**Issue by tb00** [btcsuite/btcd#687]
I updated btcd to b87723cd94ea11c29e22c4372ba4fe96886e7c83 (and applied a few unrelated local changes) which made an IDB necessary. After moving the old database out of way I pointed btcd to my backup node running Satoshi:0.11.2 (`btcd --connect a.b.c.d --nolisten --addrindex --txindex`) and let it run overnight.

In the morning I found IDB stalled at night and quickly checked the logs. After a graceful shutdown and an attempt to restart btcd it immediately exited with 'file missing', which seems to indicate leveldb corruption.

The first error happed while processing block #362911. I could track back the error to [utxoviewpoint.go#L275](https://github.com/btcsuite/btcd/blob/b87723cd94ea11c29e22c4372ba4fe96886e7c83/blockchain/utxoviewpoint.go#L275), invoked from [validate.go#L1070](https://github.com/btcsuite/btcd/blob/b87723cd94ea11c29e22c4372ba4fe96886e7c83/blockchain/validate.go#L1070). After this error, subsequent blocks have been added as orphans up to block #382320 at which point the process stalled. 

About five hours later I noticed the issue and restarted btcd. It exited with 'file missing' which I found in [ErrMissingFiles](https://github.com/btcsuite/goleveldb/blob/master/leveldb/errors/errors.go#L67), this seems to indicate leveldb corruption.

I did not have time yet to dive into the issue so I am unsure if this could be reproduced, but I backed up the whole data directory for  postmortem analysis. Below goes the relevant part of the log:
```
02:15:29 2016-04-26 [INF] BMGR: Processed 10 blocks in the last 10.76s (6493 transactions, height 362900, 2015-06-28 11:52:55 +0200 CEST)
02:15:42 2016-04-26 [INF] BMGR: Processed 7 blocks in the last 12.12s (7200 transactions, height 362907, 2015-06-28 13:48:18 +0200 CEST)
02:15:46 2016-04-26 [ERR] BMGR: Failed to process block 0000000000000000162fe85daa362cb10fea55e2d54c6273235a2c2f2771cdf0: assertion failed: view missing input 065aba8abae4684630f36cbb10deb2806684ab473c728ed3192e401e722d3b1f:1
02:15:46 2016-04-26 [INF] CHAN: Adding orphan block 00000000000000000c8caaeeda6c63498541d52fce2c7a54d375695b4983c5da with parent 0000000000000000162fe85daa362cb10fea55e2d54c6273235a2c2f2771cdf0
02:15:46 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362909
02:15:46 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362908
02:15:46 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362907
(...)
02:15:46 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362865
02:15:46 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362864
02:15:46 2016-04-26 [INF] CHAN: Adding orphan block 000000000000000014f2f8968aa3ed1ec16a166e7596e8fe3a08171294b7a84b with parent 00000000000000000c8caaeeda6c63498541d52fce2c7a54d375695b4983c5da
02:15:46 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362909
02:15:46 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362908
(...)
02:15:47 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362864
02:15:47 2016-04-26 [INF] CHAN: Adding orphan block 0000000000000000061610171c4c6beeaad57bc12d7b8a4714d51e9d9e3730d7 with parent 0000000000000000150d1362bca0202043911e21d6aa7c2a52854479e0182b4d
02:15:47 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362909
02:15:47 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362908
(...)
(...)
(...)
02:46:56 2016-04-26 [INF] CHAN: Adding orphan block 00000000000000000a8dc6ed5b133d0eb2fd6af56203e4159789b092defd8ab2 with parent 00000000000000000dfe5c861300202de68a8ce489e7360eaba3e516a9a481ed
(...)
02:46:56 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362865
02:46:56 2016-04-26 [WRN] CHAN: Lookup of known valid height failed 362864
07:32:25 2016-04-26 [INF] BTCD: Received SIGINT (Ctrl+C).  Shutting down...
07:32:25 2016-04-26 [INF] BTCD: Gracefully shutting down the server...
07:32:25 2016-04-26 [WRN] SRVR: Server shutting down
07:32:25 2016-04-26 [WRN] RPCS: RPC server shutting down
07:32:25 2016-04-26 [INF] RPCS: RPC server shutdown complete
07:32:25 2016-04-26 [INF] BMGR: Block manager shutting down
07:32:25 2016-04-26 [INF] AMGR: Address manager shutting down
07:32:25 2016-04-26 [INF] BTCD: Gracefully shutting down the database...
07:32:25 2016-04-26 [INF] SRVR: Server shutdown complete
07:32:25 2016-04-26 [INF] BTCD: Shutdown complete
07:32:38 2016-04-26 [INF] BTCD: Version 0.12.0-beta
07:32:38 2016-04-26 [INF] BTCD: Loading block database from '/srv/btcd/.btcd/data/mainnet/blocks_ffldb'
07:32:41 2016-04-26 [ERR] BTCD: file missing: file missing
07:34:06 2016-04-26 [INF] BTCD: Version 0.12.0-beta
07:34:06 2016-04-26 [INF] BTCD: Loading block database from '/srv/btcd/.btcd/data/mainnet/blocks_ffldb'
07:34:07 2016-04-26 [ERR] BTCD: file missing: file missing
(btcd --connect a.b.c.d --nolisten --debuglevel=trace)
07:38:16 2016-04-26 [INF] BTCD: Version 0.12.0-beta
07:38:16 2016-04-26 [INF] BTCD: Loading block database from '/srv/btcd/.btcd/data/mainnet/blocks_ffldb'
07:38:17 2016-04-26 [ERR] BTCD: file missing: file missing
```
Later I purged the data directory and retried IDB with a phased approach from the same node, keeping an eye on it, but this time it succeded. By phased approach I mean that I stopped btcd at around 362800, took a backup, then restarted it. My intention was to capture a consistent snapshot a bit before the problematic block so I could track this down should the error repeat.

It might have been a transient issue but the way it failed and the fact that it lead to unrecoverable database corruption is strange. I checked for, but did not find any indication of hardware errors.

Did anyone experience similar errors and subsequent corruption?

**Comment by dajohi** [comment]
Please provide a diff of your local changes.

**Comment by davecgh** [comment]
Did you update leveldb to the latest version?  There was a known bug in older versions of goleveldb which the new code tickled.  Please make sure your local copy of `$GOPATH/src/github.com/btcsuite/goleveldb` is at btcsuite/goleveldb@7834afc9e8cd15233b6c3d97e12674a31ca24602.

**Comment by davecgh** [comment]
I should also note, we recently converted btcwallet to [glide](https://github.com/Masterminds/glide) and will be doing the same for btcd.

That will help prevent outdated dependency issues like this because it works alongside the go tool's vendor directory support by downloading the deps into a locla vendor directory under btcd and checking them all out at the correct version based upon a lockfile.  At that point, `go build` and `go install` end up using the the local vendor directory instead of `$GOPATH` and results in deterministic and reproducible builds.

**Comment by tb00** [comment]
@dajohi [these are my minor changes around RPC](https://github.com/btcsuite/btcd/compare/master...tb00:json-rpc-compat) to mirror the quirks of Bitcoin Core.

@davecgh thanks for pointing this out, I was a few commits behind indeed. As I mentioned I already purged the data directory and processed IDB with success - this second time with a phased approach, I stopped and restarted btcd a few times during IDB which seems to have helped. I read up on the upstream issue, it explains why the second approach had better chances. I noticed last week you commented on #676 and shared your plan to move to glide - sounds great.
